### PR TITLE
feat: Location of client service obtained from the registry

### LINKED ISSUE #3874: Location of client service obtained from the registry
In accordance with the [service registry ADR](https://docs.edgexfoundry.org/2.1/design/adr/0018-Service-Registry/) and per the monthly architect's meeting of 1/24/22, update the bootstrapping handler of each service so that it initializes service clients using information from the registry (versus configuration file).

# 🚀 Feature Request

### Relevant Package [**REQUIRED**]
This feature request is for all EdgeX services

### Description [**REQUIRED**]
Services will use service information from the registry when initializing the service clients.  See ADR for details.

### Describe the solution you'd like
A retry loop will be needed because the client service may not always register before the requesting service.

Note: this only be addressed service startup.  If a client service changes its location at runtime, the using service will not know or be made aware of the change.

### Describe alternatives you've considered
 Have you considered any alternative solutions or workarounds?

### PR DESCRIPTION
**Pull request by lenny-intel**
All core and support services now use new common handler from go-mod-bootstrap which uses registry (if enabled) to obtain the base URL.
This handler only creates the specific Clients for those configured in the [Clients] section.
For services that don't currently uses Clients and don't have any Clients in configuiration, no Clients are created.

closes #3874

This PR is dependent on https://github.com/edgexfoundry/go-mod-bootstrap/pull/305 and is in draft mode until it is merged.

Signed-off-by: lenny <leonard.goodell@intel.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?) **N/A**
- [X] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions
Build and run Core Command locally with rest of EdgeX non-secure stack running in Docker (docker core command stopped)
Use curl or Postman to send GET to `localhost:59882/api/v2/device/all`
Verify response received with available commands for Device Virtual

## New Dependency Instructions (If applicable)
**N/A**

### CODE REVIEW

**Review by cloudxxx8 — APPROVED** [review]
LGTM

File: internal/core/metadata/main.go [hunk 1]
```diff
@@ -25,14 +25,15 @@ import (
 	pkgHandlers "github.com/edgexfoundry/edgex-go/internal/pkg/bootstrap/handlers"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/telemetry"
 
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/gorilla/mux"
+
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/core/metadata/main.go:30]: Goland IDE inserts them in this way and collapses the import section so these breaks are not readily apparent. I'll fix the one in the PR, but IMO this is not something we should care about.

[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/core/metadata/main.go]: @farshidtz , I pushed commit to sort imports.

**Comment by farshidtz** [comment]
Is this making support-notification mandatory?

I built and installed the snap, but core-metadata exits because it can't find the endpoint of support-notification. Note that support-notification is disabled by default in the snap.

```
2022-02-10T17:56:05+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:05.874265847Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:06+01:00 edgexfoundry.consul[203864]: 2022-02-10T17:56:06.542+0100 [WARN]  agent: Check is now critical: check=core-metadata
2022-02-10T17:56:06+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:06.875713036Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:07+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:07.877340453Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:08+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:08.879053007Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:09+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:09.880496307Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:10+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:10.88145789Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:11+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:11.883315881Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:12+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:12.885131722Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:13+01:00 edgexfoundry.core-metadata[204800]: level=WARN ts=2022-02-10T16:56:13.886647248Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."
2022-02-10T17:56:14+01:00 edgexfoundry.core-metadata[204800]: level=ERROR ts=2022-02-10T16:56:14.88711237Z app=core-metadata source=clients.go:60 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. Giving up"
2022-02-10T17:56:14+01:00 edgexfoundry.core-metadata[204800]: level=INFO ts=2022-02-10T16:56:14.887198802Z app=core-metadata source=database.go:135 msg="Database for V2 API disconnected"
2022-02-10T17:56:14+01:00 edgexfoundry.core-metadata[204800]: level=INFO ts=2022-02-10T16:56:14.887240109Z app=core-metadata source=config.go:455 msg="Watching for '/Writable' configuration changes has stopped"
2022-02-10T17:56:14+01:00 edgexfoundry.core-metadata[204800]: level=INFO ts=2022-02-10T16:56:14.887239789Z app=core-metadata source=secrets.go:284 msg="context cancelled, dismiss the token renewal process"
2022-02-10T17:56:14+01:00 edgexfoundry.core-metadata[204800]: level=INFO ts=2022-02-10T16:56:14.887457307Z app=core-metadata source=bootstrap.go:135 msg="Un-Registering service from the Registry"
2022-02-10T17:56:14+01:00 edgexfoundry.consul[203864]: 2022-02-10T17:56:14.899+0100 [INFO]  agent: Deregistered service: service=core-metadata
```

Works fine if I also start support-notifications:
```
sudo snap start edgexfoundry.support-notifications
sudo snap restart edgexfoundry.core-metadata
```

File: internal/core/metadata/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (farshidtz) [reply internal/core/metadata/main.go]: Thanks Lenny. You are right and this doesn't change anything in the final binaries. But then we could have the same argument for code formatting.

Apart from readability, this keeps the history cleaner. You can see that the diff now doesn't include lines that you didn't really wanna change. This also reduced merge conflicts when someone changes other parts of the code and makes other imports.

For Goland, you may able to prevent that from happening: https://stackoverflow.com/a/70668482/3041544

**Comment by lenny-intel** [comment]
> I also think the intent was that the service startup logic would only enforce the dependency on notifications if this setting was set to `true`.

Yep, see that now. Humm. this puts more complexity of having common Clients handler in bootstrap. But should be able to handle it with a parameter.

**Comment by tonyespy** [comment]
When I build the snap from your latest commit, core-metadata is now failing because it can't get the client for support-notifications.

`level=WARN ts=2022-02-10T22:37:41.529552149Z app=core-metadata source=clients.go:137 msg="unable to Get service endpoint for 'support-notifications': no matching service endpoint found. retrying..."`

**Comment by tonyespy** [comment]
OK, with the latest commit, I'm able to build the snap, core-metadata comes up and stays up, and I'm able to enable device-virtual and I'm able to use the edgex-cli to query all of the supported commands from core-command.

I'll do a more formal review of the code tomorrow.

**Review by farshidtz — APPROVED** [review]
Thanks for the fixes @lenny-intel .

I can confirm that it works now with:
- `PostDeviceChanges = false`
- `PostDeviceChanges = true` (modified in consul or override with env var) + support-notifications

File: internal/core/command/controller/http/command_test.go [hunk 2]
```diff
@@ -349,10 +349,10 @@ func TestIssueGetCommand(t *testing.T) {
 
 	dic := NewMockDIC()
 	dic.Update(di.ServiceConstructorMap{
-		bootstrapContainer.MetadataDeviceClientName: func(get di.Get) interface{} { // add v2 API MetadataDeviceClient
+		bootstrapContainer.DeviceClientName: func(get di.Get) interface{} { // add v2 API DeviceClient
 			return dcMock
 		},
-		bootstrapContainer.MetadataDeviceServiceClientName: func(get di.Get) interface{} { // add v2 API MetadataDeviceProfileClient
+		bootstrapContainer.DeviceServiceClientName: func(get di.Get) interface{} { // add v2 API DeviceProfileClient
```
Comment (tonyespy) [inline internal/core/command/controller/http/command_test.go:355]: mismatch between the code & comment on this line (comment should say `DeviceServiceClient`)

File: internal/security/secretstore/secure-messagebus.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ default:
   port: 6379
   protocol: redis
   server: localhost
-  serviceServer: http://localhost:48080
+  connectionSelector: edgex.redisMsgBus
```
Comment (tonyespy) [inline internal/security/secretstore/secure-messagebus.go:43]: How is this related to the registry changes?

**Review by tonyespy — CHANGES_REQUESTED** [review]
Overall looks good to me, although I had a few comments which may need to be addressed.

As mentioned in an earlier comment, I was finally able to build the snap and test, and all works as expected now.

File: internal/security/secretstore/secure-messagebus.go (hunk 3 above)
Comment (farshidtz) [reply internal/security/secretstore/secure-messagebus.go:43]: This and the other two snap changes come from https://github.com/edgexfoundry/edgex-go/commit/85bf44e29f7317f38ef2a818ce6085ebcdec776e merged from main into this branch. I think they will be removed from the diff on squash+merge.

File: internal/core/command/controller/http/command_test.go (hunk 2 above)
Comment (lenny-intel) [reply internal/core/command/controller/http/command_test.go]: Yep, and those comments are not useful ;-) Removed

File: internal/security/secretstore/secure-messagebus.go (hunk 3 above)
Comment (lenny-intel) [reply internal/security/secretstore/secure-messagebus.go]: @tonyespy , @farshidtz , Yep something went wrong on a rebase/merge and the PR picked up previous commits for previous PR that was already merged. Now fixed.

**Comment by lenny-intel** [comment]
@tonyespy , squashed commits to resolve issue with rebase/merge that brought in previous commits and addressed your other comments. THX!
